File: typescript/src/binary-search/binary-search.ts
// vim: noexpandtab

/**
 * Returns first index of found needle within order array, or error if needle was not found
 */
export function binarySearch<T>(array: T[], needle: T): number {
	let lo = 0;
	let hi = array.length;

	do {
		const middle_index = Math.floor(lo + (hi - lo) / 2);
		const middle_value = array[middle_index];

		if (middle_value === needle) {
			return middle_index;
		} else if (middle_value > needle) {
			hi = middle_index;
		} else {
			lo = middle_index + 1;
		}
	} while (lo < hi);

	throw new Error('Did not find needle in array');
}

File: go/binary-search/binary-search.go
package binary_search

import (
	"cmp"
	"errors"
	"math"
)

type OrderedSlice[T cmp.Ordered] []T

/**
 * Returns first index of found needle within order array, or error if needle was not found
 */
func (array OrderedSlice[T]) Binary_Search(needle T) (int, error) {
	lo := 0
	hi := len(array)

	for ok := true; ok; ok = (lo < hi) {
		middle_index := int(math.Floor(float64(lo) + (float64(hi)-float64(lo))/float64(2)))
		middle_value := array[middle_index]

		if middle_value == needle {
			return middle_index, nil
		} else if middle_value > needle {
			hi = middle_index
		} else {
			lo = middle_index + 1
		}
	}

	return int(-1), errors.New("Did not find needle in array")
}
